cuML/src/glm/preprocess.h

```diff
@@ -45,50 +47,54 @@ void preProcessData(const cumlHandle_impl& handle, math_t *input, int n_rows, in
+				stream);
 
 		Stats::mean(mu_labels, labels, 1, n_rows, false, false, stream);
-		Stats::meanCenter(labels, labels, mu_labels, 1, n_rows, false, true, stream);
+		Stats::meanCenter(labels, labels, mu_labels, 1, n_rows, false, true,
+				stream);
 
 		if (normalize) {
 			LinAlg::colNorm(norm2_input, input, n_cols, n_rows, LinAlg::L2Norm, false,
-                                        stream,
-                                        []__device__(math_t v){ return MLCommon::mySqrt(v); });
-			Matrix::matrixVectorBinaryDivSkipZero(input, norm2_input, n_rows, n_cols, false, true, stream, true);
+					stream,
+					[]__device__(math_t v) {return MLCommon::mySqrt(v);});
+			Matrix::matrixVectorBinaryDivSkipZero(input, norm2_input, n_rows,
+					n_cols, false, true, stream, true);
 		}
 	}
 
 }
 
 template<typename math_t>
-void postProcessData(const cumlHandle_impl& handle, math_t *input, int n_rows, int n_cols, math_t *labels, math_t *coef,
-		math_t *intercept, math_t *mu_input, math_t *mu_labels, math_t *norm2_input,
+void postProcessData(const cumlHandle_impl& handle, math_t *input, int n_rows,
+		int n_cols, math_t *labels, math_t *coef, math_t *intercept,
+		math_t *mu_input, math_t *mu_labels, math_t *norm2_input,
 		bool fit_intercept, bool normalize, cudaStream_t stream) {
-    auto cublas_handle = handle.getCublasHandle();
-    auto cusolver_handle = handle.getcusolverDnHandle();
+
+	auto cublas_handle = handle.getCublasHandle();
+	auto cusolver_handle = handle.getcusolverDnHandle();
 
 	ASSERT(n_cols > 0,
 			"Parameter n_cols: number of columns cannot be less than one");
 	ASSERT(n_rows > 1,
 			"Parameter n_rows: number of rows cannot be less than two");
 
-	math_t *d_intercept;
-	allocate(d_intercept, 1);
+	auto allocator = handle.getDeviceAllocator();
+	device_buffer<math_t> d_intercept(allocator, stream, 1);
 
 	if (normalize) {
-            Matrix::matrixVectorBinaryMult(input, norm2_input, n_rows, n_cols, false, true, stream);
-            Matrix::matrixVectorBinaryDivSkipZero(coef, norm2_input, 1, n_cols,
-                                                    false, true, stream, true);
+		Matrix::matrixVectorBinaryMult(input, norm2_input, n_rows, n_cols,
+				false, true, stream);
+		Matrix::matrixVectorBinaryDivSkipZero(coef, norm2_input, 1, n_cols,
+				false, true, stream, true);
 	}
 
-	LinAlg::gemm(mu_input, 1, n_cols, coef, d_intercept, 1, 1,
-		    CUBLAS_OP_N, CUBLAS_OP_N, cublas_handle, stream);
+	LinAlg::gemm(mu_input, 1, n_cols, coef, d_intercept.data(), 1, 1,
+			CUBLAS_OP_N, CUBLAS_OP_N, cublas_handle, stream);
 
-	LinAlg::subtract(d_intercept, mu_labels, d_intercept, 1, stream);
-	updateHost(intercept, d_intercept, 1, stream);
-        CUDA_CHECK(cudaStreamSynchronize(stream));
-	if (d_intercept != NULL)
-		cudaFree(d_intercept);
+	LinAlg::subtract(d_intercept.data(), mu_labels, d_intercept.data(), 1,
+			stream);
+	updateHost(intercept, d_intercept.data(), 1, stream);
```

Comment:
Does the caller of `postProcessData` expect that the correct value of `intercept` can be read after `postProcessData` has returned? If yes a stream synchronize is needed after this.

Reply:
We call deviceSync in python level after each fit function. So, that should make sure intercept value is correct.

Reply:
I think we should consider that cuML might be used from a C++ level. I.e. the C++ API should have clearly and well defined semantics. Does that make sense to you?

Reply:
Makes sense. This functions was originally designed for cython level. Otherwise, wouldn't copy intercept to host. Anyways, will add stream sync. Thanks for the comment.

Reply:
@jirikraus  is the function "waitOnInternalStreams" in cumlhandle that I can use for stream sync?

Reply:
@jirikraus  Ok, added the stream sync. Please approve this PR if everything looks good to you.